sympy/core/exprtools.py

```diff
@@ -1414,156 +1288,614 @@ def factor_nc(expr):
+                return _NCMonomial(self.coeff / other.coeff, (), self.generators, degree=0)
+            return None
+        if other.degree == 0:
+            return _NCMonomial(self.coeff / other.coeff, self.term, self.generators, self.degree)
+        other_len = len(other.term)
+        if other.term[:-1] != self.term[:other_len - 1]:
+            return None
+        if other.term[-1][0] != self.term[other_len - 1][0]:
+            return None
+        if other.term[-1][1] > self.term[other_len - 1][1]:
+            return None
+        if other.term[-1][1] == self.term[other_len - 1][1]:
+            return _NCMonomial(
+                self.coeff / other.coeff,
+                self.term[other_len:],
+                self.generators,
+                degree=self.degree - other.degree)
+        else:
+            pw = self.term[other_len - 1][1] - other.term[-1][1]
+            return _NCMonomial(
+                self.coeff / other.coeff,
+                ((self.term[other_len - 1][0], pw),) + self.term[other_len:],
+                self.generators,
+                degree=self.degree - other.degree)
+
+    def rdivide(self, other):
+        """Return the monomial m such that self == m * other
+
+        If no such monomial exists, return None
+        """
+        if other.degree > self.degree:
+            return None
+        if other.degree == self.degree:
+            if self.term == other.term:
+                return _NCMonomial(self.coeff / other.coeff, (), self.generators, degree=0)
+            return None
+        if other.degree == 0:
+            return _NCMonomial(self.coeff / other.coeff, self.term, self.generators, self.degree)
+        other_len = len(other.term)
+        if other_len > 1 and other.term[1:] != self.term[-other_len - 1:]:
+            return None
+        if other.term[0][0] != self.term[-other_len][0]:
+            return None
+        if other.term[0][1] > self.term[-other_len][1]:
+            return None
+        if other.term[0][1] == self.term[-other_len][1]:
+            return _NCMonomial(
+                self.coeff / other.coeff,
+                self.term[:-other_len],
+                self.generators,
+                degree=self.degree - other.degree)
+        else:
+            pw = self.term[-other_len][1] - other.term[0][1]
+            return _NCMonomial(
+                self.coeff / other.coeff,
+                self.term[:-other_len] + ((self.term[-other_len][0], pw),),
+                self.generators,
+                degree=self.degree - other.degree)
+
+
+class _NCPoly:
+    """Class that represents a polynomial with noncommutative generators
+
+    It is used only for factoring expressions with noncommutative terms and does not have the
+    same interface as the Poly class (which supports only commutative terms).
+
+    A few functions (add, subtract) update an instance in-place, unlike most sympy classes."""
+    def __init__(self, expr, generators=None, extensions=None):
```

Comment:
I also would like to exclude `extensions` from the public, if it is not obviously used.